tests/test_tensornet_tf.py

```diff
@@ -0,0 +1,579 @@
+
+        def circuit(x):
+            qml.Rot(0.3, x, -0.2, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        node = qml.QNode(circuit, tensornet_tf_device_2_wires)
+
+        with pytest.raises(ValueError, match="Parameter indices must be unique."):
+            node.jacobian(0.5, which=[0, 0])
+
+    def test_indices_nonexistant(self, tensornet_tf_device_2_wires):
+        """Tests that QNode.jacobian properly raises an error if the
+           jacobian is requested for non-existant parameters."""
+
+        def circuit(x):
+            qml.Rot(0.3, x, -0.2, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        node = qml.QNode(circuit, tensornet_tf_device_2_wires)
+
+        with pytest.raises(ValueError, match="Tried to compute the gradient wrt"):
+            node.jacobian(0.5, which=[0, 6])
+
+        with pytest.raises(ValueError, match="Tried to compute the gradient wrt"):
+            node.jacobian(0.5, which=[1, -1])
+
+    def test_unknown_method(self, tensornet_tf_device_2_wires):
+        """Tests that QNode.jacobian properly raises an error if the
+           gradient method is unknown."""
+
+        def circuit(x):
+            qml.Rot(0.3, x, -0.2, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        node = qml.QNode(circuit, tensornet_tf_device_2_wires)
+
+        with pytest.raises(ValueError, match="Unknown gradient method"):
+            node.jacobian(0.5, method="unknown")
+
+    def test_wrong_order_in_finite_difference(self, tensornet_tf_device_2_wires):
+        """Tests that QNode.jacobian properly raises an error if finite
+           differences are attempted with wrong order."""
+
+        def circuit(x):
+            qml.Rot(0.3, x, -0.2, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        node = qml.QNode(circuit, tensornet_tf_device_2_wires)
+
+        with pytest.raises(ValueError, match="Order must be 1 or 2"):
+            node.jacobian(0.5, method="F", order=3)
+
+    # Parameters obtained using the following three lines of code:
+    # ALLOWED_DIFF_METHODS = ("best", "parameter-shift", "finite-diff")
+    # ALLOWED_INTERFACE = ("autograd", "numpy", "torch", "tf", None)
+    # diff_method_interface_products = product(ALLOWED_DIFF_METHODS, ALLOWED_INTERFACE)
+
+    @pytest.mark.parametrize("diff_method, interface", [
+        ('best', 'autograd'),
+        ('best', 'numpy'),
+        ('best', 'torch'),
+        ('best', 'tf'),
+        ('best', None),
+        ('parameter-shift', 'autograd'),
+        ('parameter-shift', 'numpy'),
+        ('parameter-shift', 'torch'),
+        ('parameter-shift', 'tf'),
+        ('parameter-shift', None),
+        ('finite-diff', 'autograd'),
+        ('finite-diff', 'numpy'),
+        ('finite-diff', 'torch'),
+        ('finite-diff', 'tf'),
+        ('finite-diff', None),
+    ])
+    def test_jacobian(self, tensornet_tf_device_3_wires, diff_method, interface, tol):
+
```

Comment:
docstring would be useful here. In particular, mention that the test is comparing against `default.qubit`